Postpone Serato timing offset correction until audio signal properties are known

Holzhaus opened the pull request:
This is an attempt to solve the timing offset issue discussed here:

https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Serato.20cue.20import/near/195095023

~~This doesn't work, for some reason the detected timing offset is always `0`, but I don't now why.~~

Holzhaus committed:
track/serato: Determine timing offsets when audio signal properties are known

Sometimes, cue offsets depend on audio signal properties such as the
track's sample rate. This is the case for cues imported from Serato when
the track is imported with the MAD decoder. Before, the timing offsets
were determined before the sample rate is known. This change allows
postponing the offset calculation until the audio signal properties are
known.

Zulip Discussion: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Serato.20cue.20import/near/195090899

fixup

Be-ing commented on src/track/serato/tags.cpp:
```diff
@@ -129,38 +133,34 @@ double SeratoTags::findTimingOffsetMillis(const QString& filePath) {
+            timingOffset = -24;
+        } else if (signalInfo.getSampleRate() == 44000) {
+            timingOffset = -26;
         }
```
add an else branch to warn that we don't know the timing offset for other sample rates

Be-ing commented on src/track/serato/tags.cpp:
```diff
@@ -129,38 +133,34 @@ double SeratoTags::findTimingOffsetMillis(const QString& filePath) {
+        if (signalInfo.getSampleRate() == 48000) {
+            timingOffset = -24;
+        } else if (signalInfo.getSampleRate() == 44000) {
+            timingOffset = -26;
```
comment that this is an estimate and tracks will unpredictably vary within ~1 ms

uklotzde replied:
Probably 44100 and not 44000

Holzhaus commented:
> This doesn't work, for some reason the detected timing offset is always `0`, but I don't now why.

I suppose this has something to do with the Qt container classes. Do they copy the object and silently convert them to the parent class?

Holzhaus commented:
@uklotzde Is this a worthwhile approach? The other idea I had was to let `importCuePoints` take a function pointer to a function that returns the timing offset. Or we just open the file in `findTimingOffsetMillis`. They are all hacky, I'm just not sure how else we could solve that problem.

Holzhaus commented:
> > This doesn't work, for some reason the detected timing offset is always `0`, but I don't now why.
> 
> I suppose this has something to do with the Qt container classes. Do they copy the object and silently convert them to the parent class?

Yes they do. I have no idea how to fix this.

uklotzde commented:
> > This doesn't work, for some reason the detected timing offset is always `0`, but I don't now why.
> 
> I suppose this has something to do with the Qt container classes. Do they copy the object and silently convert them to the parent class?

Unless you allocate dynamically and store a pointer you won't get polymorphic behavior. The containers store just what they have been specified with, i.e. the base class. Slicing occurs when adding a derived class to the container. RTTI only works for references and pointers, but you store plain values in the container that are all of the same type both at compile time and at runtime.

Holzhaus commented:
@uklotzde Thanks, I switched to shared pointers and it works now.

Be-ing commented on src/track/serato/tags.cpp:
```diff
@@ -112,13 +115,14 @@ double SeratoTags::findTimingOffsetMillis(const QString& filePath) {
     // PR for more detailed information:
     // https://github.com/mixxxdj/mixxx/pull/2119
 
+    qWarning() << "sampleRate" << signalInfo.getSampleRate();
```
remove

Be-ing commented on src/track/track.cpp:
```diff
@@ -893,8 +893,20 @@ void Track::importCueInfosMarkDirtyAndUnlock(
+        // Fix wrong timing offsets for cue points
+        double timingOffset = pCueInfo->getTimingOffsetMillis(
+                getLocation(), m_streamInfo->getSignalInfo());
+        qWarning() << "TIMING OFFSET" << timingOffset;
```
remove

daschuer commented:
Please add a brief outline how this code is intended to work. I have not followed the discussion, so I am a bit lost. 

Please correct me if I am wrong:

Depending on the sample rate or encoder the cues can be shifted. 
This means the cues stored in the database are no longer the positions from Mixxx, they can be imported from everywhere. 
Along with the original cue points Mixxx has other type of cues, Serato / Recordbox ...

I think this can work, at least we do not mass with the external Data and can tweak the ofset calculation code whenever we find out more.

Holzhaus commented:
> The `CueInfo` class should not know anything about offset calculations that are only needed during import. Instead I suggest to revert all changes in `CueInfo` and instead create a new `CueImport` class that contains `CueInfo` as a member and provides the polymorphic offset calculation.

Yes, good idea. I was worried about this as well, but making a separate importer class didn't occur to me. 

> Please add a brief outline how this code is intended to work. I have not followed the discussion, so I am a bit lost.

OK, sorry. We found a new way to determine cue offsets more accurately then just by comparing the waveforms visually:

https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Serato.20cue.20import/near/195090899

When looking at the results, we found out that the Serato cue offsets with the MAD decoder depend on the sample rate. Right now, we determine the offsets before the sample rate is known. This PR aims to postpone that until we do know the sample rate. 

> Please correct me if I am wrong:
>
>Depending on the sample rate or encoder the cues can be shifted.
Correct. 

> This means the cues stored in the database are no longer the positions from Mixxx, they can be imported from everywhere.

No, this has nothing to do with the database. 
We already calculate the frame offsets from the millisecond offsets at import time *as soon as we know the sample rate*. This PR only postpones the cue offset calculation until the sample rate is known as well. The position values we store in the DB are not affected.

uklotzde commented:
It is bad practice to overload a class with functions that it doesn't need intrinsically. Adding a new, dedicated class just for this purpose is the preferred way. The new type also indicates that the offset correction has not happened yet, while `CueInfo` is lacking this information.

uklotzde commented on src/track/cueinfo.h:
```diff
@@ -54,6 +55,16 @@ class CueInfo {
 
+    /// This functions can be overridden in subclasses to allow audio signal
+    /// dependent timing offset correction.
+    virtual double getTimingOffsetMillis(
```
Now `CueInfo` needs a virtual destructor. But we don't want that ;)

Be-ing commented on src/track/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,63 @@
+/// cueinfoimporter.cpp
+/// Created 2020-04-24 by Jan Holthuis
+
```
This is already in the Git history. Do we really need to add it?

Holzhaus replied:
I saw it in some files in the mixxx codebase. I can remove it if you want.

Be-ing replied:
please do

uklotzde commented on src/track/serato/tags.cpp:
```diff
@@ -224,7 +224,7 @@ QList<CueInfo> SeratoTags::getCues(const QString& filePath) const {
     };
 
-    return cueMap.values();
+    return CueInfoImporterPointer(new SeratoCueInfoImporter(cueMap.values()));
```
make_shared

uklotzde commented on src/track/serato/tags.h:
```diff
@@ -41,7 +44,7 @@ class SeratoTags final {
     }
 
-    QList<CueInfo> getCues(const QString& filePath) const;
+    CueInfoImporterPointer getCueInfoImporter() const;
```
"get" doesn't fit. You are actually importing the cue points here, i.e. importCueInfos(). Only the result is evaluated lazily, like a promise/future. You have to provide the sample rate as the key to unlock the results.

uklotzde commented on src/track/cueinfoimporter.h:
```diff
@@ -0,0 +1,37 @@
+    void append(const CueInfo& cueInfo);
+    void append(const QList<CueInfo>& cueInfos);
+
+    QList<CueInfo> getCueInfos() const;
```
Do you need both functions?

In my opinion the importer should strictly only do 2 things:
- Collect new cue infos
- Convert the collected cue infos once, i.e. becoming empty after the conversion

Holzhaus commented:
Sorry for force-pushing, I wanted to rephrase the original commit message as it didn't fit properly. You can see the code change I made here: https://github.com/mixxxdj/mixxx/compare/d8a19f590265fee659f29513b33a783c4302a3fe..0a391aa5f80cd16469b08170884f6b3ef7cecc94

Holzhaus commented:
Mac and Ubuntu complain about `std::shared_ptr`. Anyone knows how to fix this?

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into serato-offset-correction

Holzhaus committed:
track/cueinfoimporter: Add <memory> include for std::shared_ptr

Some compilers (e.g. on our OSX and Ubuntu 18.04 CI builds) fail when
this is missing:

    In file included from src/track/cueinfoimporter.cpp:1:0:
    src/track/cueinfoimporter.h:34:14: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
     typedef std::shared_ptr<CueInfoImporter> CueInfoImporterPointer;
                  ^~~~~~~~~~

uklotzde commented:
@Holzhaus Depending on how serious the merge conflicts are you could also consider rebasing.

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into serato-offset-correction

Holzhaus commented:
> @Holzhaus Depending on how serious the merge conflicts are you could also consider rebasing.

Thanks, but I was almost finished resolving the conflicts. I think this does the trick. Please review.

Be-ing commented on src/track/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,79 @@
+    // Do not calculate offset if we don't have any cues to import
+    if (m_cueInfos.isEmpty()) {
+        cueInfos.append(m_cueInfos);
+        m_cueInfos.clear();
```
`m_cueInfos` is already empty, so is this line needed?

Holzhaus replied:
woops, removed.

Be-ing commented on src/track/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    // Do not calculate offset if we don't have any cues to import
+    if (m_cueInfos.isEmpty()) {
+        cueInfos.append(m_cueInfos);
```
`cueInfos = m_cueInfos`? Or maybe simply `return m_cueInfos`?

Holzhaus replied:
@uklotzde requested that the CueInfoImporter object becomes empty after retrieving the cues, so we can't simply return that 
list: https://github.com/mixxxdj/mixxx/pull/2707#discussion_r414911370
The alternative would be to add a `clear()` method and emptying the internal list from outside. I'd also be okay with that.

uklotzde replied:
The intention is to perform this transformation exactly once. The code should reflect this use case and prevent any invalid usage if possible. This is achieved by consuming the input data during the conversion.

If you use a functional language with immutable data structures or Rust with its *move semantics* this pattern is very common ;)

Be-ing replied:
How about just `return cueInfos` with the default constructed `QList<CueInfo>`? Appending an empty list to an empty list doesn't do anything, does it?

Be-ing commented on src/track/serato/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,13 @@
+double SeratoCueInfoImporter::getTimingOffsetMillis(
+        const QString& filePath,
+        const audio::SignalInfo& signalInfo) const {
+    return SeratoTags::findTimingOffsetMillis(filePath, signalInfo);
```
How about deleting `SeratoTags::findingTimingOffsetMillis` and moving its implementation here?

Holzhaus replied:
I thought about moving the function, but we'll need that function for export, too (and export won't use the CurInfoImporter class).

Be-ing replied:
Okay, then add a comment here.

Be-ing commented:
I like this new implementation with the separate CueInfoImporter class. Good idea to keep this timing correction outside of the tag parsing code.

Holzhaus committed:
track/cueinfoimporter: Remove some unnecessary code if cueInfos is empty

Holzhaus committed:
track/serato/cueinfoimporter: Add comment for getTimingOffsetMillis()

Be-ing commented on src/track/serato/tags.cpp:
```diff
@@ -129,38 +132,40 @@ double SeratoTags::findTimingOffsetMillis(const QString& filePath) {
+            timingOffset = -26;
+            break;
+        default:
+            qWarning() << "Unknown timing offset for sample rate" << signalInfo.getSampleRate();
```
```suggestion
            qWarning() << "Unknown timing offset for Serato tags with sample rate" << signalInfo.getSampleRate();
```

Holzhaus committed:
track/serato/tags: Improve warning message text

Co-Authored-By: Be <be.0@gmx.com>

uklotzde commented on src/track/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,79 @@
+QList<CueInfo> CueInfoImporter::getCueInfosWithCorrectTiming(
+        const QString& filePath,
+        const audio::SignalInfo& signalInfo) {
+    QList<CueInfo> cueInfos;
```
We can pick the internal list and don't need to create a new list.

Try this:
```
QList<CueInfo> cueInfos(std::move(m_cueInfos));
DEBUG_ASSERT(m_cueInfos.isEmpty());
```

The debug assertion simply documents our intention.

Holzhaus replied:
Is this okay, too? https://github.com/mixxxdj/mixxx/pull/2707/files#diff-a90e67d47a0490f3d58383410d1b5a2eR48

uklotzde replied:
We still don't need this separate list and can import m_cueInfos directly.

Holzhaus replied:
I don't understand, what separate list?

```cpp
// This does involve a separate list
QList<CueInfo> cueInfos(std::move(m_cueInfos));
DEBUG_ASSERT(m_cueInfos.isEmpty());
return cueInfos;

// This doesn't, right?
return std::move(m_cueInfos);
```

uklotzde replied:
https://github.com/uklotzde/mixxx/commit/fd1a544e670eae7d9a4af66c974e39020e8a258c

uklotzde replied:
Sorry, forgot to remove newCueInfo

uklotzde replied:
https://github.com/uklotzde/mixxx/commit/749790baeba370b583a0af254d4ddb5efd96c28b

uklotzde commented on src/track/cueinfoimporter.cpp:
```diff
@@ -0,0 +1,79 @@
+    return m_cueInfos.isEmpty();
+}
+
+QList<CueInfo> CueInfoImporter::getCueInfosWithCorrectTiming(
```
Use a verb other than "get", e.g. "importCueInfosWithCorrectTiming". This is an "importer" because it's purpose is to "import".

uklotzde commented on src/track/cueinfoimporter.h:
```diff
@@ -0,0 +1,38 @@
+/// signal info (channel number, sample rate, bitrate) is known.
+class CueInfoImporter {
+  public:
+    CueInfoImporter();
```
= default;

uklotzde commented on src/track/cueinfoimporter.h:
```diff
@@ -0,0 +1,38 @@
+class CueInfoImporter {
+  public:
+    CueInfoImporter();
+    CueInfoImporter(const QList<CueInfo>& cueInfos);
```
explicit

uklotzde commented on src/track/serato/cueinfoimporter.h:
```diff
@@ -0,0 +1,15 @@
+#pragma once
+#include "track/cueinfoimporter.h"
```
newline after pragma

uklotzde commented on src/track/serato/cueinfoimporter.h:
```diff
@@ -0,0 +1,15 @@
+
+class SeratoCueInfoImporter : public CueInfoImporter {
+  public:
+    using CueInfoImporter::CueInfoImporter;
```
~SeratoCueInfoImporter() override = default;

Holzhaus replied:
Shoudn't it just use the default destructor from the parent class?

uklotzde replied:
Best practice, just to ensure that the destructor of the base class is virtual. Safe C++ is verbose.

uklotzde commented on src/track/cueinfoimporter.h:
```diff
@@ -0,0 +1,38 @@
+    /// Returns audio signal dependent timing offset correction.
+    /// The default implementation just returns 0, but this can be overridden
+    /// in subclasses.
+    virtual double getTimingOffsetMillis(
```
"getTimingOffsetMillis" -> "guessTimingOffsetMillis"

uklotzde commented on src/track/serato/tags.h:
```diff
@@ -19,7 +21,8 @@ class SeratoTags final {
     static RgbColor storedToDisplayedSeratoDJProCueColor(RgbColor color);
     static RgbColor displayedToStoredSeratoDJProCueColor(RgbColor color);
-    static double findTimingOffsetMillis(const QString& filePath);
+    static double findTimingOffsetMillis(
```
guessTimingOffsetMillis

uklotzde commented on src/track/track.h:
```diff
@@ -267,8 +268,8 @@ class Track : public QObject {
     /// import and returns the corresponding status.
-    CueImportStatus importCueInfos(
-            const QList<mixxx::CueInfo>& cueInfos = {});
+    CueImportStatus importCueInfosFromImporter(
```
The suffix "FromImporter" is awkward to read and doesn't provide any value. Just importCueInfos().

Holzhaus committed:
track/cueinfoimporter: Use default/explicit constructors

Holzhaus committed:
track: Rename methods for offset retrieval to guessTimingOffsetMillis()

Holzhaus committed:
track/cueinfoimporter: Add newline after #pragma once

Holzhaus committed:
track/cueinfoimporter: Rename method to importCueInfosWithCorrectTiming

Holzhaus committed:
track/cueinfoimporter: Use std::move instead of copying the CueInfo list

Holzhaus committed:
track/track: Remove "FromImporter" suffix from method name

Holzhaus committed:
track/cueinfoimporter: Add virtual default destructor

Be-ing commented:
There is a merge conflict now.

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into serato-offset-correction

uklotzde commented on src/track/serato/tags.cpp:
```diff
@@ -114,11 +117,11 @@ double SeratoTags::findTimingOffsetMillis(const QString& filePath) {
 
     double timingOffset = 0;
     if (filePath.toLower().endsWith(".mp3")) {
```
We should not come up with a 3rd method for determining the file type. The whole file path is not necessary if we only use need and use the file type for determining the offset.

There are different options. One can be found in trackmetadata_file.h, including a type-safe enum:
```
FileType getFileTypeFromFileName(const QString& fileName);
```

uklotzde replied:
Another option would be `QString SoundSource::getFileExtensionFromUrl(QUrl url);` that returns the string that the track uses for the *file type* property. But I guess that the metadata enum fits better here.

Holzhaus replied:
Done.

Holzhaus committed:
track/serato/tags: Use getFileTypeFromFileName() to detect file type

uklotzde commented on src/track/track.cpp:
```diff
@@ -952,8 +961,9 @@ void Track::importPendingCueInfosMarkDirtyAndUnlock(
-    for (const auto& cueInfo : m_importCueInfosPending) {
+    cuePoints.reserve(m_pCueInfoImporterPending->size());
+    for (const auto& cueInfo : m_pCueInfoImporterPending->importCueInfosWithCorrectTiming(
+                 getLocation(), m_streamInfo->getSignalInfo())) {
```
I just noticed that we should better pass `const QString& fileType = m_record.getFileType()` here instead of `const QString& filePath = getLocation()`. The Importer does not need the full `filePath`, only the `fileType`.

Unfortunately the file type is by convention just a lowercase string and not type-safe. We could introduce a constant `kFileTypeMp3` or function `bool isFileTypeMp3(...)` to avoid literal strings in the Serato code.

Holzhaus replied:
> I just noticed that we should better pass `const QString& fileType = m_record.getFileType()` here instead of `const QString& filePath = getLocation()`. The Importer does not need the full `filePath`, only the `fileType`.

We still need the full file path for mp3guessenc. Currently, the CoreAudio soundsource uses that for guessing the offsets.

Be-ing commented:
Ready for merge?

uklotzde commented:
@Be Wait, let me try something.

uklotzde committed:
Import cue points in-place

Be-ing commented:
Now ready for merge?

uklotzde commented:
I haven't performed a manual test. But I think so. LGTM

Holzhaus commented:
Merge?
